Some atom work

ploeh opened the pull request:
Leaving for a meeting, so this may stop a bit abruptly... Back in a few hours :)

BTW. Please don't use the "automatic merge" button. We need to establish proper rebasing procedures to handle pull requests. Ping me if you need help with the mechanics of it.

Commit:
Added some properties to TestEvent,

to make sure that some data is actually being serialized.

Commit:
Added AtomEntryFileWriter,

which will be used to implement ISyndicationItemWriter by writing files in
the Atom format.

Commit:
Added an ID to the item.

This ID is essentially the 'changeset ID', as outlined here:
http://seabites.wordpress.com/2011/12/07/your-eventstream-is-a-linked-list/

Commit:
Added an author to the entry.

This is mainly to satisfy the Atom specification, that states that if the
entry is to be used as a stand-alone document (which it may be in this
implementation), it MUST have at least one author, and the author MUST
have at least a name.
At the moment I've simply hard-coded the name to "Grean", but in a more
advanced scenario, we could consider pulling the name from a set of
claims, so that we actually have a sort of audit trail of the originators
of each changeset.
For that reason, I also made the test code quite lax in its requirements.

Commit:
Added UuidIri struct,

intended to be a Value Object modeling a uuid IRI. The motivation is that
I've already identifed three places in the code base where it refers to
the hard-coded string "urn:uuid", which tells me that I can invoke the
Rule of Three. I foresee more uuid parsing and matching in my immediate
future.

Commit:
Converted UuidIri to Guid.

This is a lossless conversion, so an implicit conversion is appropriate.

Commit:
Converted Guid to UuidIti -

another lossless conversion.

Commit:
Ensured correct Value Object identity.

Apparently, a struct wrapping a single other struct correctly implements
Equals, even without overriding Equals.
That means that the two tests I've added here automatically pass, so to
make them fail I actually tried to override Equals and incorrectly
implement them.
The tests still provide regression protection, so should be kept in place.

Commit:
Added UuidIri.TryParse.

Commit:
Refactored implementation of TryParse.

Commit:
Extracted prefix to a const.

Commit:
Checked for null in TryParse.

Commit:
Refactored Resemblance to use new UuidIri type.

sgryt commented:
This one
-        [Theory, AutoAtomData]
-        public void TryParseIncorrectlyPrefixedStringReturnsCorrectResult(
-            Guid guid)
-        {
-            var inccorectlyPrefixed = "this.is:not:right" + guid;
  +
-            UuidIri actual;
-            bool couldParse = UuidIri.TryParse(inccorectlyPrefixed, out actual);
  +
-            Assert.False(couldParse, "TryParse should fail.");
-            Assert.Equal(default(UuidIri), actual);
-        }

seems like it could be made stronger if you parametrize the incorrect prefix as, say, 'string thisIsNotRight' instead of hardcoding the string value 'this.is:not:right' in the test ?

Other than that, I am a bit lost due to my lack of knowledge of the ATOM spec :)

On Apr 10, 2013, at 1:05 PM, Mark Seemann notifications@github.com wrote:

> Leaving for a meeting, so this may stop a bit abruptly... Back in a few hours :)
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/GreanTech/AtomEventStore some-atom-work
> Or view, comment on, or merge it at:
> 
>   https://github.com/GreanTech/AtomEventStore/pull/1
> 
> Commit Summary
> 
> Added some properties to TestEvent,
> Added AtomEntryFileWriter,
> Added an ID to the item.
> Added an author to the entry.
> Added UuidIri struct,
> Converted UuidIri to Guid.
> Converted Guid to UuidIti -
> Ensured correct Value Object identity.
> Added UuidIri.TryParse.
> Refactored implementation of TryParse.
> Extracted prefix to a const.
> Checked for null in TryParse.
> Refactored Resemblance to use new UuidIri type.
> File Changes
> 
> A AtomEventStore.UnitTests/AtomEntryFileWriterTests.cs (20)
> M AtomEventStore.UnitTests/AtomEventStore.UnitTests.csproj (2)
> M AtomEventStore.UnitTests/SyndicationItemResemblance.cs (18)
> M AtomEventStore.UnitTests/TestEvent.cs (3)
> A AtomEventStore.UnitTests/UuidIriTests.cs (115)
> A AtomEventStore/AtomEntryFileWriter.cs (15)
> M AtomEventStore/AtomEventStore.csproj (2)
> M AtomEventStore/SyndicationStore.cs (2)
> A AtomEventStore/UuidIri.cs (48)
> Patch Links:
> 
> https://github.com/GreanTech/AtomEventStore/pull/1.patch
> https://github.com/GreanTech/AtomEventStore/pull/1.diff

Commit:
Strengthened a test

by turning a hard-coded string representing an incorrect prefix into an
anonymous variable representing an incorrect prefix.
A Guard Assertion documents the assumption that the incorrectPrefix
variable is NOT equal to the correct prefix.

ploeh commented:
Good point. Corrected.
